joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/passes/typerelations/PointerAssignmentPass.scala

```diff
@@ -0,0 +1,143 @@
+package io.joern.x2cpg.passes.typerelations
+
+import io.shiftleft.codepropertygraph.Cpg
+import io.shiftleft.codepropertygraph.generated.nodes._
+import io.shiftleft.codepropertygraph.generated.{EdgeTypes, Operators}
+import io.shiftleft.passes.ForkJoinParallelCpgPass
+import io.shiftleft.semanticcpg.language._
+
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+/** Creates an intraprocedural pointer assignment graph. This provides may point to information for each identifier or
+  * field identifier.
+  */
+class PointerAssignmentPass(cpg: Cpg) extends ForkJoinParallelCpgPass[Method](cpg) {
+
+  override def generateParts(): Array[Method] = cpg.method.toArray
+
+  override def runOnPart(builder: DiffGraphBuilder, method: Method): Unit = {
+    val allocSites = method.ast.isCall.name(Operators.alloc, Operators.arrayInitializer).l
+    if (allocSites.nonEmpty) {
+      val assignmentGraph = buildAssignmentGraph(method)
+      val pointsToGraph   = buildPointsToGraph(assignmentGraph)
+      // All references to nodes in the method body should now have their
+      // "may point to" edge added to an allocation site if it is in the map
+      method.ast
+        .collect {
+          case x: Identifier      => x
+          case y: FieldIdentifier => y
+        }
+        .map(new VarInCtx(_))
+        .foreach { varDecl =>
+          pointsToGraph.get(varDecl) match {
+            case Some(tgtNodes) =>
+              tgtNodes.foreach { tgt => builder.addEdge(varDecl.node, tgt.node, EdgeTypes.POINTS_TO) }
+            case None =>
+          }
+        }
+    }
+  }
+
+  /** Builds a graph where identifiers on the left of assignment statements are mapped to nodes on the right hand side.
+    */
+  private def buildAssignmentGraph(m: Method): Map[VarInCtx, Set[VarInCtx]] = {
+    val assignmentGraph               = mutable.HashMap.empty[VarInCtx, Set[VarInCtx]]
+    var worklist: ListBuffer[CfgNode] = mutable.ListBuffer.from(m.ast.isExpression.assignment.l)
+
+    while (worklist.nonEmpty) {
+      val assignment = worklist.head
+      worklist = worklist.tail
+      // Extract LHS and RHS of each expression
+      val allocLhs: Set[VarInCtx] = assignment.astChildren
+        .orderLte(1)
+        .collect {
+          case x: Identifier => VarInCtx(x)
+          case x: Call if x.methodFullName == Operators.fieldAccess =>
+            x.astChildren.collect { case x: FieldIdentifier => VarInCtx(x) }.head
```

Comment:
This is super problematic. An assignment
```
foo.fieldname = y;
```
will now seemingly overwrite the `fieldname` field of all objects that have a field named `fieldname` and forget about `foo`.

Also, `alloclhs` is an Option, no need to make a Set out of it.